File: Pemrograman-berorientasi-objek/Praktikum/Praktikum-1/Kompleks.cpp
#include <iostream>
#include "Kompleks.h"

using namespace std;

int Kompleks::n_kompleks = 0;

Kompleks::Kompleks() {
	n_kompleks += 1;
	this->real = 0;
	this->imaginer = 0;
}

Kompleks::Kompleks(int real, int imaginer) {
	n_kompleks += 1;
	this->real = real;
	this->imaginer = imaginer;
}

int Kompleks::GetReal() const {
	return real;
}

int Kompleks::GetImaginer() const {
	return imaginer;
}

void Kompleks::SetReal(int r) {
	real = r;
}

void Kompleks::SetImaginer(int i) {
	imaginer = i;
}

Kompleks operator+(const Kompleks& a, const Kompleks& b) {
	Kompleks temp(a.GetReal() + b.GetReal(), a.GetImaginer() + b.GetImaginer());
	return temp;
}

Kompleks operator-(const Kompleks& a, const Kompleks& b) {
	Kompleks temp(a.GetReal() - b.GetReal(), a.GetImaginer() - b.GetImaginer());
	return temp;
}

Kompleks operator*(const Kompleks& a, const Kompleks& b) {
	// ac - bd , bc + ad
	int r = a.GetReal() * b.GetReal() - a.GetImaginer() * b.GetImaginer();
	int i = a.GetImaginer() * b.GetReal() + a.GetReal() * b.GetImaginer();
	Kompleks temp(r, i);
	return temp;
}

Kompleks operator*(const Kompleks& a, const int b) {
	Kompleks temp(a.GetReal() * b, a.GetImaginer() * b);
	return temp;
}

Kompleks operator*(const int b, const Kompleks& a) {
	Kompleks temp(a.GetReal() * b, a.GetImaginer() * b);
	return temp;
}

int Kompleks::CountKompleksInstance() {
	return n_kompleks;
}

void Kompleks::Print() {
	if (imaginer < 0) {
		cout << real << imaginer << "i" << endl;
	} else {
		cout  << real << "+" << imaginer << "i" << endl;		
	}
}

File: Pemrograman-berorientasi-objek/Praktikum/Responsi-4/Kompleks.java

class Kompleks {
    private static int n_kompleks = 0;
    private int real;
    private int imaginer;

    public Kompleks() {
        this.real = 0;
        this.imaginer = 0;
        n_kompleks++;
    }

    public Kompleks(int real, int imaginer) {
        this.real = real;
        this.imaginer = imaginer;
        n_kompleks++;
    }

    public int getReal() {
        return this.real;
    }

    public int getImaginer() {
        return this.imaginer;
    }

    public void setReal(int real) {
        this.real = real;
    }

    public void setImaginer(int imaginer) {
        this.imaginer = imaginer;
    }

    public Kompleks plus(Kompleks b) {
        int real = this.real + b.getReal();
        int imaginer = this.imaginer + b.getImaginer();
        return new Kompleks(real, imaginer);
    }

    public Kompleks minus(Kompleks b) {
        int real = this.real - b.getReal();
        int imaginer = this.imaginer - b.getImaginer();
        return new Kompleks(real, imaginer);
    }

    public Kompleks multiply(Kompleks b) {
        int real = this.real * b.getReal() - this.imaginer * b.getImaginer();
        int imaginer = this.imaginer * b.getReal() + this.real * b.getImaginer();
        return new Kompleks(real, imaginer);
    }

    public Kompleks multiply(int b) {
        int real = this.real * b;
        int imaginer = this.imaginer * b;
        return new Kompleks(real, imaginer);
    }

    public static int countKompleksInstance() {
        return n_kompleks;
    }

    private void printIfNotZero(int a, String b) {
        if (a != 0) {
            System.out.print(a + b);
        }
    }

    public void print() {
        if (this.getReal() == 0 && this.getImaginer() == 0) {
            System.out.println(0);
        } else {
            if (this.getImaginer() < 0) {
                this.printIfNotZero(this.getReal(), "");
                this.printIfNotZero(this.getImaginer(), "i");
                System.out.println();
            } else {
                this.printIfNotZero(this.getReal(), "");
                if (this.getReal() != 0 && this.getImaginer() != 0) {
                    System.out.print("+");
                }
                this.printIfNotZero(this.getImaginer(), "i");
                System.out.println();

            }
        }
    }

    public static void main(String[] args) {
        Kompleks a = new Kompleks(-5, 4);
        Kompleks b = new Kompleks(-4, -5);
        System.out.println(a.getReal());
        a.print();
        b.print();
    }

}